cmd/xsDeploy.go

```diff
@@ -0,0 +1,424 @@
+	}
+
+	action, err := ValueOfAction(XsDeployOptions.Action)
+	if err != nil {
+		fmt.Printf("Extracting action failed: %v\n", err)
+		return err
+	}
+
+	if mode == Deploy && action != None {
+		return errors.New(fmt.Sprintf("Cannot perform action '%s' in mode '%s'. Only action '%s' is allowed.", action, mode, None))
+	}
+
+	log.Entry().Debugf("Mode: '%s', Action: '%s'", mode, action)
+
+	performLogin := mode == Deploy || (mode == BGDeploy && !(action == Resume || action == Abort))
+	performLogout := mode == Deploy || (mode == BGDeploy && action != None)
+	log.Entry().Debugf("performLogin: %t, performLogout: %t", performLogin, performLogout)
+
+	if action == None && !fExists(XsDeployOptions.MtaPath) {
+		return errors.New(fmt.Sprintf("Deployable '%s' does not exist", XsDeployOptions.MtaPath))
+	}
+
+	if action != None && len(XsDeployOptions.DeploymentID) == 0 {
+		return errors.New(fmt.Sprintf("deploymentID was not provided. This is required for action '%s'.", action))
+	}
+
+	prOut, pwOut := io.Pipe()
+	prErr, pwErr := io.Pipe()
+
+	s.Stdout(pwOut)
+	s.Stderr(pwErr)
+
+	var e, o string
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		buf := new(bytes.Buffer)
+		r := io.TeeReader(prOut, os.Stderr)
+		io.Copy(buf, r)
+		o = buf.String()
+		wg.Done()
+	}()
+
+	go func() {
+		buf := new(bytes.Buffer)
+		r := io.TeeReader(prErr, os.Stderr)
+		io.Copy(buf, r)
+		e = buf.String()
+		wg.Done()
+	}()
+
+	var loginErr error
+
+	xsSessionFile := ".xsconfig"
+	if len(XsDeployOptions.XsSessionFile) > 0 {
+		xsSessionFile = XsDeployOptions.XsSessionFile
+	}
+
+	if performLogin {
+		loginErr = xsLogin(XsDeployOptions, s)
+		if loginErr == nil {
+			err = copyFileFromHomeToPwd(xsSessionFile, fCopy)
+		}
+	}
+
+	if loginErr == nil && err == nil {
+
+		if !fExists(xsSessionFile) {
+			return fmt.Errorf("xs session file does not exist (%s)", xsSessionFile)
+		}
+
+		copyFileFromPwdToHome(xsSessionFile, fCopy)
+
+		switch action {
+		case Resume, Abort, Retry:
+			err = complete(mode, action, XsDeployOptions.DeploymentID, s)
+		default:
+			err = deploy(mode, XsDeployOptions, s)
+		}
+	}
+
+	if loginErr == nil && (performLogout || err != nil) {
+		if logoutErr := xsLogout(XsDeployOptions, s); logoutErr != nil {
+			if err == nil {
+				err = logoutErr
+			}
+		} else {
+
+			// we delete the xs session file from workspace. From home directory it is deleted by the
+			// xs command itself.
+			if e := fRemove(xsSessionFile); e != nil {
+				err = e
+			}
+			log.Entry().Debugf("xs session file '%s' has been deleted from workspace", xsSessionFile)
+		}
+	} else {
+		if loginErr != nil {
+			log.Entry().Info("Logout skipped since login did not succeed.")
+		} else if !performLogout {
+			log.Entry().Info("Logout skipped in order to be able to resume or abort later")
+		}
+	}
+
+	if err == nil {
+		err = loginErr
+	}
+
+	if err != nil {
+		if _, e := os.Stat(fmt.Sprintf("%s/%s", os.Getenv("HOME"), ".xs_logs")); !os.IsNotExist(e) {
+			s.RunShell("/bin/bash",
```

Comment:
why do we need a shell call for that. Shouldn't we handle this directly in golang

Reply:
Was some c&p from the groovy code. Can be done on the pure go level, of course.

Reply:
I agree that doing that via go lang file facilities in in general better compared to doing this via shell. But one point why this is like it is now: In tests we can for now catch that in the corresponding shell runner mock. Currently we do not have the corresponding file system mocks. That means we would really access the file system. Having dependencies to the file system in tests is not that cool since the status of the file system (which files are present and which are not present) affects the tests --> maybe we should agree on some pattern for mocking file system access. But again: I agree, in general that should be done on go-level not via `cat` on shell level.

Reply:
done, now we serve the log files on pure go level. No shell used for that anymore.